---
title: "LocusXcanR: An R Shiny App to Visualize TWAS Results Integrated with Other 'Omics Studies"
date: "06/08/2020"
author: "Amanda L. Tapia"
output: 
  html_document:
    toc: true
    toc_float: true
    number_sections: true
abstract: |
  Lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum
  lorem ipsum lorem ipsum lorem ipsum lorem ipsum lorem ipsum

vignette: |
  %\VignetteIndexEntry{LocusXcanR: An R Shiny App to Visualize TWAS Results Integrated with Other 'Omics Studies}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(LocusXcanR)
```

# Introduction

Use sections (and sub-sections) to demarcate meaningful steps in the process, e.g. “Preprocessing”, “Quality control”, “Main analysis”, “Plotting results”, “Diagnostics”, etc. You can specify toc: true (and toc_float: true) in the output section of the preamble in order to add a Table of Contents (and make it floating, that is, unfold/fold and follow as the user scrolls down). See the Rmarkdown cheat sheet for instructions on adding options to the preamble.
Most of the code chunks should be evaluated, it’s ok to add a few eval=FALSE code chunks for demonstration, but vignettes where none of the code chunks are evaluated are not as helpful in my opinion, as the output that a user gets when running the code may differ from what happens in the non-evaluated code chunks, which may even generate errors if they are out-of-date!

```{r, include=FALSE}
dir <- system.file("extdata", package="LocusXcanR")
rdir <- system.file("data", package="LocusXcanR")

twas_ds <- read.table(file.path(dir, "twas_ds.txt"), header=T, stringsAsFactors = F)
load(file.path(rdir,"pred_exp_corr.Rdata"))
```

# LocusXcanR R function
LocusXcanR <- function(twas_result,weight_tbl,study_name="",pred_exp_corr,conditional_present=FALSE,multiple_tissues=FALSE,
                      known_variants,known_gwas,db_genes,all_gwas,ld_gwas,ref_expr_name="",head_details="",method_details="",
                      primary_tissue,meta_present=FALSE){}


# Example Data


##  TWAS results dataset  
The data set <code>twas_ds</code> requires the following variables:  
<ul>
  <li><code>tissue</code>:</li>
  <li><code>locus</code>: integer, value indicating the locus number</li>
  <li><code>locstart</code>: integer, value indicating the genomic start position of the locus</li>
  <li><code>locstop</code>: integer, value indicating the genomic stop position of the locus</li>
  <li><code>chr</code>: integer, chromosome number</li>
  <li><code>index</code>: character, the index (most significant) gene within the locus</li>
  <li><code>pheno</code>:</li>
  <li><code>genestart</code>:</li>
  <li><code>genestop</code>:</li>
  <li><code>p</code>:</li>
  <li><code>genename</code>:</li>
</ul>

Additional variables may also be included in the dataset. Any additional variables will be ignored
when plotting figures but will be printed in data tables.

Here is an example of <code>twas_ds</code>:
```{r}
knitr::kable(twas_ds[1:5,])
```

## GWAS results dataset

Here is an example of <code>known_variants</code>:
```{r}
knitr::kable(twas_ds[1:5,])
```

## Predicted expression correlation dataset
Name of correlation matrix object must be saved as capital <code>M</code>.

Here is an example of the <code>M</code> object loaded by the <code>pred_exp_corr</code> parameter.
```{r}
knitr::kable(M[1:5,1:5])
```


# Frequently Asked Questions

# Acknowledgements
list people who helped in writing the software

# References
list relevant literature
See the preamble above for how to add a BibTex file (and see Rmarkdown cheat sheet for how to add references to an Rmarkdown file).
